### ISSUE TITLE
Problem with immutable call

### ISSUE THREAD

**Issue by box-system-design** [javers/javers#816]
Hi,

In the Sets class, we can found this:

```
public static <E> Set<E> xor(Set<E> first, Set<E> second) {
        **Set<E> xor = difference(first, second);**
        **xor.addAll(difference(second, first));**
        return xor;
}

/**
* null args are allowed
*/
public static <E> Set<E> difference(Set<E> first, Set<E> second) {
        if (first == null || first.size() == 0) {
            return Collections.emptySet();
        }

        if (second == null || second.size() == 0) {
            return first;
        }

        Set<E> difference = new HashSet<>(first);
        difference.removeAll(second);
        return difference;
}
```

In my case, the first line in bold `Set<E> xor = difference(first, second);` go into this if 
```
if (first == null || first.size() == 0) {
    return Collections.emptySet();
}
```
 and return an immutable set. 

The second line in bold `xor.addAll(difference(second, first));` go into this if 
```
if (second == null || second.size() == 0) {
     return first;
}
```
and return one element. 

The addAll just after generate an UnsupportedOperationException.

Here is the stacktrace:

```
2019-03-27 17:12:01.361 ERROR 5156 --- [http-nio-11000-exec-7] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.UnsupportedOperationException] with root cause

java.lang.UnsupportedOperationException: null
    at java.util.AbstractCollection.add(AbstractCollection.java:262)
    at java.util.AbstractCollection.addAll(AbstractCollection.java:344)
    at org.javers.common.collections.Sets.xor(Sets.java:49)
    at org.javers.core.metamodel.object.CdoSnapshotState.differentValues(CdoSnapshotState.java:114)
    at org.javers.core.metamodel.object.CdoSnapshotBuilder.build(CdoSnapshotBuilder.java:76)
    at org.javers.core.snapshot.SnapshotFactory.createUpdate(SnapshotFactory.java:53)
    at org.javers.core.snapshot.ChangedCdoSnapshotsFactory.createUpdateSnapshot(ChangedCdoSnapshotsFactory.java:65)
    at org.javers.core.snapshot.ChangedCdoSnapshotsFactory.createSnapshot(ChangedCdoSnapshotsFactory.java:53)
    at org.javers.core.snapshot.ChangedCdoSnapshotsFactory.create(ChangedCdoSnapshotsFactory.java:41)
    at org.javers.core.commit.CommitFactory.createCommit(CommitFactory.java:88)
    at org.javers.core.commit.CommitFactory.create(CommitFactory.java:75)
    at org.javers.core.JaversCore.commit(JaversCore.java:82)
    at org.javers.spring.auditable.aspect.springdata.AbstractSpringAuditableRepositoryAspect.lambda$new$0(AbstractSpringAuditableRepositoryAspect.java:33)
    at org.javers.spring.auditable.aspect.springdata.AbstractSpringAuditableRepositoryAspect.lambda$onVersionEvent$3(AbstractSpringAuditableRepositoryAspect.java:52)
```

I think that the first fragment must return a new HashSet or something similar, no?

Thanks in advance for your help.

**Comment by bartoszwalacik** [comment]
Sets class is not a part of Javers public API, please dont use it.

**Comment by box-system-design** [comment]
Why do you close my issue? Could you take the time to read my problem please?

I don't use directly this Sets class. I'm in a spring boot app. I activate javers with annotation on m'y repository class and when I want to save something the versioning crash on this Sets class because it tries toi add an élément to an immutable collections.

So please take the time to look.

**Comment by bartoszwalacik** [comment]
That's another story, please read and follow [Guidelines for Bug Reporting](https://github.com/javers/javers/blob/master/CONTRIBUTING.md#guidelines-for-bug-reporting)
Example test cases: https://github.com/javers/javers/tree/master/javers-core/src/test/groovy/org/javers/core/cases

**Comment by bartoszwalacik** [comment]
this is a duplicate to https://github.com/javers/javers/issues/815
will be fixed in soon

**Comment by bartoszwalacik** [comment]
fixed in 5.3.4
